ons-spark/spark-concepts/partitions.md

```diff
@@ -0,0 +1,941 @@
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/CalYear=2010
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/CalYear=2011
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/CalYear=2012
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/CalYear=2013
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/CalYear=2014
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/CalYear=2015
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/CalYear=2016
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/CalYear=2017
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/CalYear=2018
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/CalYear=2019
+file:///home/cdsw/ons-spark/checkpoints/rescue_by_year.parquet/_SUCCESS
+```
+On the right of the ouput above you will see there is one directory for each `CalYear`. So to import a subset of the data we can use the specific path for that year or filter the data in Spark and let Spark work out which folders to look for. 
+
+Finally, we will delete these files to clean up the file system.
+````{tabs}
+```{code-tab} py
+cmd = f"hdfs dfs -ls -C {repartition_path}"
+p = subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, universal_newlines=True)
+```
+
+```{code-tab} r R
+
+cmd <- paste0("hdfs dfs -rm -r -skipTrash ", repartition_path)
+system(cmd)
+
+```
+````
+## How should I partition my data?
+
+The short answer is- don't worry about it too much!
+
+Let's flip the question around. What are the consequences of getting the number of partitions "wrong"? Most of the time- nothing. You could spend hours trying to optimise the number of partitions and gain a $x2$ increase in processing speed, but it's likely there are other things you can do that will gain $x10$ or $x100$ increase in processing speed. 
+
+As the computer scientist Donald Knuth once wrote, 
+
+>Premature optimization is the root of all evil (or at least most of it) in programming
+>
+>    -> *Computer Programming as an Art (1974)*
+
+More importantly it's good practice to be aware of the size of the DataFrame and the number of partitions as this will help to avoid obvious issues or over-partitioning, under-partitioning and highly skewed processes.
+
+A more accurate answer depends on a variety of factors including: the size of the data, data types, distributions within the data, type of processing and other properties defined in the `SparkSession`. Here are some tips for quick wins:
+
+- as suggested above, the first step is getting the code right. Only look to optimise if you're running into performance issues
+- decreasing the `spark.sql.shuffle.partitions` parameter is sensible for smaller datasets
+- if you google it, you might find the optimum size for a Spark partitions is 100-200MB. In practice, this is impossible to keep track of in an analysis script, but it gives you an idea of what is considered big or small
```

Comment:
I prefer to not promote brands so would replace this with "if you use a search engine..." or similar

I think Google needs a capital letter if you leave it in - although not entirely sure, has it entered the language?

Reply:
good point, done